### PR TITLE
Add optional IAP support to Cloud Run services.

### PR DESCRIPTION
**Pull request by wlynch**
cc @jonjohnsonjr @imjasonh

### CODE REVIEW

File: modules/serverless-gclb/variables.tf [hunk 1]
```diff
@@ -34,3 +34,12 @@ variable "notification_channels" {
   type        = list(string)
   default     = []
 }
+
+variable "iap" {
+  description = "IAP configuration for the load balancer."
+  type = object({
+    oauth2_client_id     = string
+    oauth2_client_secret = string
```
Comment (imjasonh) [inline modules/serverless-gclb/variables.tf:42]: Is this something we should mark as sensitive?

**Comment by imjasonh** [comment]
Do you have an example of how this is used, both from the IAP config side, and from the app side to get the currently IAP'ed user's identity and/or token?

File: modules/serverless-gclb/variables.tf (hunk 1 above)
Comment (wlynch) [reply modules/serverless-gclb/variables.tf:42]: The GCP resource already marks this sensitive in the generated output. I also don't know if we can mark it as sensitive here (https://discuss.hashicorp.com/t/is-it-possible-to-mark-an-attribute-of-an-object-as-sensitive/24649)

```
resource "google_compute_backend_service" "public-services" {
      + connection_draining_timeout_sec = 300
      + creation_timestamp              = (known after apply)
      + fingerprint                     = (known after apply)
      + generated_id                    = (known after apply)
      + id                              = (known after apply)
      + load_balancing_scheme           = "EXTERNAL"
      + name                            = "nami-frontend"
      + port_name                       = (known after apply)
      + project                         = "chainguard-nami"
      + protocol                        = (known after apply)
      + self_link                       = (known after apply)
      + session_affinity                = (known after apply)
      + timeout_sec                     = (known after apply)

      + backend {
          + balancing_mode               = "UTILIZATION"
          + capacity_scaler              = 1
          + group                        = (known after apply)
          + max_connections              = (known after apply)
          + max_connections_per_endpoint = (known after apply)
          + max_connections_per_instance = (known after apply)
          + max_rate                     = (known after apply)
          + max_rate_per_endpoint        = (known after apply)
          + max_rate_per_instance        = (known after apply)
          + max_utilization              = (known after apply)
            # (1 unchanged attribute hidden)
        }

      + iap {
          + oauth2_client_id            = "123526589617-dccnftqc6snrcpa4o4anvm286bsamoe7.apps.googleusercontent.com"
          + oauth2_client_secret        = (sensitive value)
          + oauth2_client_secret_sha256 = (sensitive value)
        }
    }
```

**Comment by wlynch** [comment]
https://cloud.google.com/iap/docs/enabling-cloud-run has good documentation on this.

This bit of terraform is really configuring the GCLB with the IAP app's oauth creds so it knows how to auth.

There is an extra step where you need to grant the IAP SA cloud run invoker, but this can't easily be done within these modules:

```terraform
resource "google_project_service" "iap" {
  project = var.project_id
  service = "iap.googleapis.com"

  timeouts {
    create = "30m"
    update = "40m"
  }

  disable_on_destroy = false
}

resource "google_project_service_identity" "iap_sa" {
  provider = google-beta

  project = var.project_id
  service = "iap.googleapis.com"
}

resource "google_cloud_run_service_iam_member" "invoker" {
  for_each   = module.networking.regional-networks
  depends_on = [google_project_service.iap]
  project    = var.project_id
  location   = each.key
  service    = var.name
  role       = "roles/run.invoker"
  member     = "serviceAccount:${google_project_service_identity.iap_sa.email}"
}
```

from there you can grant access by granting users to the IAP via https://cloud.google.com/iap/docs/managing-access#add_access

**Comment by wlynch** [comment]
On the app side, there a bunch of headers that IAP provides. a few useful ones;

- `X-Goog-Authenticated-User-Email` - authenticated user's email
- `X-Goog-Iap-Jwt-Assertion` - federated user oidc token

  ```json
      {
          "aud": "/projects/123526589617/global/backendServices/842338515430489131",
          "azp": "/projects/123526589617/global/backendServices/842338515430489131",
          "email": "billy@chainguard.dev",
          "exp": 1719507640,
          "hd": "chainguard.dev",
          "iat": 1719507040,
          "identity_source": "GOOGLE",
          "iss": "https://cloud.google.com/iap",
          "sub": "accounts.google.com:104847474376810186341"
      }
    ```

- `X-Serverless-Authorization` - oidc token of IAP -> app

  ```json
  {
    "aud": "/projects/123526589617/locations/us-east4/services/foo",
    "azp": "107617952505205565908",
    "email": "service-123526589617@gcp-sa-iap.iam.gserviceaccount.com",
    "email_verified": true,
    "exp": 1719436271,
    "iat": 1719432671,
    "iss": "https://accounts.google.com",
    "sub": "107617952505205565908"
  }    
  ```

**Review by imjasonh — APPROVED** [review]

**Pull request merged**